### PR TITLE
feat(tracer): Add DD_TAGS environment variable support

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
Specifying the environment variable `DD_TAGS` will allow users to specify tags to be attached to every span.

### CODE REVIEW

File: ddtrace/utils/formats.py [hunk 1]
```diff
@@ -89,3 +90,27 @@ def flatten_dict(d, sep=".", prefix=""):
...
+    """
+    tags = {}
+    if not envstr:
+        return tags
+
+    for tag in envstr.split(","):
+        try:
+            key, value = tag.split(":", 1)
+        except ValueError as e:
+            e.args = (e.args or tuple()) + (tag,)
```
Comment (Kyle-Verhoog) [inline ddtrace/utils/formats.py]: I don't like this.. but I guess the alternative is subclassing `ValueError` and adding an argument/attribute for `tag` 😕.

File: ddtrace/profiling/exporter/http.py [hunk 2]
```diff
@@ -110,6 +111,12 @@ def _get_tags(service):
         }
         user_tags = os.getenv("DD_PROFILING_TAGS")
         if user_tags:
+            try:
+                parse_env_tags(user_tags)
```
Comment (jd) [inline ddtrace/profiling/exporter/http.py]: Not using the result?

File: ddtrace/utils/formats.py (hunk 1 above)
Comment (jd) [reply ddtrace/utils/formats.py]: You can just raise a new `ValueError`, no?

File: ddtrace/utils/formats.py [hunk 3]
```diff
@@ -89,3 +90,27 @@ def flatten_dict(d, sep=".", prefix=""):
...
+            raise
+        else:
+            if isinstance(value, six.text_type):
+                value = value.encode("utf-8")
```
Comment (jd) [inline ddtrace/utils/formats.py]: Use `six.ensure_binary`

File: ddtrace/profiling/exporter/http.py [hunk 4]
```diff
@@ -110,15 +111,10 @@ def _get_tags(service):
...
-                    key, value = tag.split(":", 1)
-                except ValueError:
-                    LOG.error("Malformed tag in DD_PROFILING_TAGS: %s", tag)
-                else:
-                    if isinstance(value, six.text_type):
-                        value = value.encode("utf-8")
-                    tags[key] = value
+            parsed_tags, err_tags = parse_env_tags(user_tags)
+            if err_tags:
+                LOG.error("Malformed tags in DD_PROFILING_TAGS: %s", ",".join(err_tags))
```
Comment (jd) [inline ddtrace/profiling/exporter/http.py]: if you always end up logging, log from `parse_env_tags` directly?

**Comment by majorgreys** [comment]
Is this something we can make use of also in parsing `DATADOG_PATCH_MODULES`?

https://github.com/datadog/dd-trace-py/blob/master/ddtrace/bootstrap/sitecustomize.py#L53-L67

**Comment by Kyle-Verhoog** [comment]
Good catch! I guess we can rename this function to `parse_key_value_str` or similar. Yay less code duplication 😄

**Comment by Kyle-Verhoog** [comment]
@jd we keep them as strings AFAIK, the code for parsing the tags was moved from the profiler (it seems like tags in the profiler are stored as byte strings?)

File: ddtrace/bootstrap/sitecustomize.py [hunk 5]
```diff
@@ -54,29 +54,15 @@ def update_patched_modules():
...
-        if len(patch.split(":")) != 2:
-            log.debug("skipping malformed patch instruction")
-            continue
 
-        module, should_patch = patch.split(":")
-        if should_patch.lower() not in ["true", "false"]:
+    modules = parse_tags_str(modules_to_patch)
+    for module, should_patch in modules.items():
+        should_patch = should_patch.lower()
+        if should_patch not in ["true", "false"]:
```
Comment (jd) [inline ddtrace/bootstrap/sitecustomize.py:61]: `ddtrace.utils.formats.asbool` ?

**Comment by Kyle-Verhoog** [comment]
I _was_ calling `ensure_binary` because the code I based this off of was from the profiler 😛. I since removed the call since in the tracer we don't store tags as bytes (although we probably should...).

File: ddtrace/bootstrap/sitecustomize.py (hunk 5 above)
Comment (Kyle-Verhoog) [reply ddtrace/bootstrap/sitecustomize.py:61]: yeah we may as well do it here too. It wasn't originally an `asbool` but I see no reason to not use it here. 😄

File: ddtrace/utils/formats.py [hunk 6]
```diff
@@ -89,3 +93,31 @@ def flatten_dict(d, sep=".", prefix=""):
...
+    :param tags_str: A string of the above form to parse tags from.
+    :return: A dict containing the tags that were parsed.
+    """
+    parsed_tags = {}
+    if not tags_str:
+        return parsed_tags
+
+    for tag in tags_str.split(","):
+        try:
+            key, value = tag.split(":", 1)
```
Comment (brettlangdon) [inline ddtrace/utils/formats.py:113]: you can also use `key, _, value = tag.partition(":")` will basically promise that it won't fail to parse.

File: ddtrace/utils/formats.py [hunk 7]
```diff
@@ -89,3 +93,31 @@ def flatten_dict(d, sep=".", prefix=""):
...
+
+            # Reject empty keys
+            if key == "":
```
Comment (brettlangdon) [inline ddtrace/utils/formats.py:116]: could also do `if not key:`

File: tests/test_utils.py [hunk 8]
```diff
@@ -101,3 +102,32 @@ def test_flatten_dict(self):
         d = dict(A=1, B=2, C=dict(A=3, B=4, C=dict(A=5, B=6)))
         e = dict(A=1, B=2, C_A=3, C_B=4, C_C_A=5, C_C_B=6)
         self.assertEquals(flatten_dict(d, sep='_'), e)
+
+    def test_parse_env_tags(self):
```
Comment (brettlangdon) [inline tests/test_utils.py:106]: pytest parametrized tests are great for this kind of stuff.

as an example: https://github.com/DataDog/dd-trace-py/blob/a6653c383d23a8e5740a8c79411d3fe352fe8467/tests/test_span.py#L444-L465

File: tests/test_utils.py [hunk 9]
```diff
@@ -101,3 +102,32 @@ def test_flatten_dict(self):
...
+        assert tags == dict(key="val")
+
+        tags = parse_tags_str("key:val,key2:val2")
+        assert tags == dict(key="val", key2="val2")
+
+        tags = parse_tags_str("key:val,key2:val2,key3:1234.23")
+        assert tags == dict(key="val", key2="val2", key3="1234.23")
+
+        parsed_tags = parse_tags_str("key:,key3:val1,")
+        assert parsed_tags == dict(key="", key3="val1")
```
Comment (brettlangdon) [inline tests/test_utils.py:117]: interesting, do we want empty values?

File: tests/test_utils.py [hunk 10]
```diff
@@ -101,3 +102,32 @@ def test_flatten_dict(self):
         d = dict(A=1, B=2, C=dict(A=3, B=4, C=dict(A=5, B=6)))
         e = dict(A=1, B=2, C_A=3, C_B=4, C_C_A=5, C_C_B=6)
         self.assertEquals(flatten_dict(d, sep='_'), e)
+
+    def test_parse_env_tags(self):
+        tags = parse_tags_str("key:val")
```
Comment (brettlangdon) [inline tests/test_utils.py:107]: Could add:

```
parse_tags_str("")
parse_tags_str(",")
parse_tags_str(":,:")
```

File: tests/test_utils.py (hunk 9 above)
Comment (Kyle-Verhoog) [reply tests/test_utils.py:117]: Yeah I played with the idea of disallowing it... Keys I think make definite sense for disallowing, values... maybe?

Comment (brettlangdon) [reply tests/test_utils.py:117]: yeah, I don't think values makes sense, it'll just added bytes sent for just the key, and I don't think dogstatsd tags can have an empty value either

File: tests/test_utils.py (hunk 8 above)
Comment (brettlangdon) [reply tests/test_utils.py:106]: (not a requirement, just a thought)

File: ddtrace/profiling/exporter/http.py [hunk 11]
```diff
@@ -147,15 +144,7 @@ def _get_tags(service):
 
         user_tags = os.getenv("DD_PROFILING_TAGS")
```
Comment (jd) [inline ddtrace/profiling/exporter/http.py:145]: I'm also wondering if we should add `DD_TAGS` here, no?

File: ddtrace/utils/formats.py [hunk 12]
```diff
@@ -89,3 +93,31 @@ def flatten_dict(d, sep=".", prefix=""):
...
+    if not tags_str:
+        return parsed_tags
+
+    for tag in tags_str.split(","):
+        try:
+            key, value = tag.split(":", 1)
+
+            # Reject empty keys or values
+            if key == "" or value == "":
+                raise ValueError
```
Comment (jd) [inline ddtrace/utils/formats.py:117]: I'm pretty sure you can have empty values for tags in metrics. Are we sure it's not the case for spans?

Comment (brettlangdon) [reply ddtrace/utils/formats.py:117]: https://docs.datadoghq.com/tagging/#defining-tags

Comment (brettlangdon) [reply ddtrace/utils/formats.py:117]: > Note: A tag cannot end with a colon, for example tag:

File: ddtrace/profiling/exporter/http.py (hunk 11 above)
Comment (brettlangdon) [reply ddtrace/profiling/exporter/http.py:145]: probably, share the same tags between APM and profiling?

File: ddtrace/profiling/exporter/http.py [hunk 13]
```diff
@@ -146,17 +143,9 @@ def _get_tags(service):
         if env:
             tags["env"] = env
 
-        user_tags = os.getenv("DD_PROFILING_TAGS")
+        user_tags = os.environ.get("DD_TAGS", os.environ.get("DD_PROFILING_TAGS"))
```
Comment (jd) [inline ddtrace/profiling/exporter/http.py:146]: I'm not sure we should ignore `DD_PROFILING_TAGS` if `DD_TAGS` is there. I'd see them at being supplementary, where `DD_PROFILING_TAGS` would override the tags set in `DD_TAGS` if there's an overlap.

File: ddtrace/utils/formats.py (hunk 12 above)
Comment (jd) [reply ddtrace/utils/formats.py:117]: Do we want to validate that here too?
It'd be better to do so if the agent is going to drop tags or traces. Having an error raised is easier to debug.

File: ddtrace/profiling/exporter/http.py (hunk 13 above)
Comment (Kyle-Verhoog) [reply ddtrace/profiling/exporter/http.py:146]: :+1: what do you think about conflicting keys? Which should get precedence?

Comment (Kyle-Verhoog) [reply ddtrace/profiling/exporter/http.py:146]: also, do we envision using `DD_PROFILING_TAGS` long-term? Or is this just for the short-term?

Comment (jd) [reply ddtrace/profiling/exporter/http.py:146]: `DD_PROFILING_TAGS` should be used to add and possibly override tags from `DD_TAGS`.

I don't know if we're going to keep long term. I guess we can, it's a cool feature to have — worst case it's just useless.

File: ddtrace/utils/formats.py (hunk 12 above)
Comment (Kyle-Verhoog) [reply ddtrace/utils/formats.py:117]: We should probably be validating on `span.set_tag`. I think I'll open up a separate PR for validating tags. Might make more sense to do it all in a consolidated place.

Comment (jd) [reply ddtrace/utils/formats.py]: agreed

**Pull request merged**